[VPlan] Hook IR blocks into VPlan during skeleton creation (NFC)

fhahn opened the pull request:
As a first step to move towards modeling the full skeleton in VPlan, start by wrapping IR blocks created during legacy skeleton creation in VPIRBasicBlocks and hook them into the VPlan. This means the skeleton CFG is represented in VPlan, just before execute. This allows moving parts of skeleton creation into recipes in the VPBBs gradually.

Note that this allows retiring some manual DT updates, as this will be handled automatically during VPlan execution.

fhahn committed:
[VPlan] Hook IR blocks into VPlan during skeleton creation (NFC)

As a first step to move towards modeling the full skeleton in VPlan,
start by wrapping IR blocks created during legacy skeleton creation in
VPIRBasicBlocks and hook them into the VPlan. This means the skeleton
CFG is represented in VPlan, just before execute. This allows moving
parts of skeleton creation into recipes in the VPBBs gradually.

Note that this allows retiring some manual DT updates, as this will be
handled automatically during VPlan execution.

ayalz reviewed (commented):
First set of comments and/or potential follow-ups.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn committed:
!fixup address comments, thanks!

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks-tmp

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2426,6 +2426,26 @@ InnerLoopVectorizer::getOrCreateVectorTripCount(BasicBlock *InsertBlock) {
+  VPBlockBase *ScalarPH = Plan.getScalarPreheader();
+  VPBlockBase *PredVPB = VectorPH->getSinglePredecessor();
+  VPBlockUtils::disconnectBlocks(Plan.getEntry(), VectorPH);
+  VPBlockUtils::connectBlocks(PredVPB, ScalarPH);
```
Updated, thanks! The new API is a bit more difficult to get right, but more efficient.

fhahn replied:
Added in 95eeae195e608797314d71f7327e638a98764471

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2426,6 +2426,26 @@ InnerLoopVectorizer::getOrCreateVectorTripCount(BasicBlock *InsertBlock) {
+  VPBlockBase *VectorPH = Plan.getVectorPreheader();
+  VPBlockBase *ScalarPH = Plan.getScalarPreheader();
+  VPBlockBase *PredVPB = VectorPH->getSinglePredecessor();
+  VPBlockUtils::disconnectBlocks(Plan.getEntry(), VectorPH);
```
Done, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2426,6 +2426,26 @@ InnerLoopVectorizer::getOrCreateVectorTripCount(BasicBlock *InsertBlock) {
+  VPIRBasicBlock *CheckVPIRBB = VPIRBasicBlock::fromBasicBlock(CheckIRBB);
+  VPBlockUtils::connectBlocks(PredVPB, CheckVPIRBB);
+  VPBlockUtils::connectBlocks(CheckVPIRBB, ScalarPH);
+  VPBlockUtils::connectBlocks(CheckVPIRBB, VectorPH);
```
Updated for now to use the new connect variant but left the everything else as is for now. Will check if there are other places where the new helper could be used.

fhahn replied:
Added in ccb40b0b7a51a0619acc8a6b479b86ff28a19e5b

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2511,13 +2531,14 @@ void InnerLoopVectorizer::emitIterationCountCheck(BasicBlock *Bypass) {
          "TC check is expected to dominate Bypass");
```
Yes, this is to check the DT updates from SplitBlock I think

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2511,13 +2531,14 @@ void InnerLoopVectorizer::emitIterationCountCheck(BasicBlock *Bypass) {
          "TC check is expected to dominate Bypass");
 
   // Update dominator for Bypass & LoopExit (if needed).
```
Removed, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2511,13 +2531,14 @@ void InnerLoopVectorizer::emitIterationCountCheck(BasicBlock *Bypass) {
   ReplaceInstWithInst(TCCheckBlock->getTerminator(), &BI);
   LoopBypassBlocks.push_back(TCCheckBlock);
+
+  connectScalarPreheaderInVPlan(Plan);
```
Added a TODO for now, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2511,13 +2531,14 @@ void InnerLoopVectorizer::emitIterationCountCheck(BasicBlock *Bypass) {
     setBranchWeights(BI, MinItersBypassWeights, /*IsExpected=*/false);
   ReplaceInstWithInst(TCCheckBlock->getTerminator(), &BI);
   LoopBypassBlocks.push_back(TCCheckBlock);
+
```
There are a few places that assume the scalar PH has a single predecessor, which would need to be updated. Could pull this in here or adjust as follow-up?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7877,6 +7899,12 @@ EpilogueVectorizerMainLoop::emitIterationCountCheck(BasicBlock *Bypass,
+  VPBlockBase *VectorPH = Plan.getVectorPreheader();
+  VPBlockBase *PredVPB = VectorPH->getSinglePredecessor();
+  if (PredVPB->getNumSuccessors() == 1)
+    connectScalarPreheaderInVPlan(Plan);
```
There are a few places that assume the scalar PH has a single predecessor, which would need to be updated. Could pull this in here or adjust as follow-up?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7877,6 +7899,12 @@ EpilogueVectorizerMainLoop::emitIterationCountCheck(BasicBlock *Bypass,
+  VPBlockBase *PredVPB = VectorPH->getSinglePredecessor();
+  if (PredVPB->getNumSuccessors() == 1)
+    connectScalarPreheaderInVPlan(Plan);
+  else
```
as in having the entry connect directly to TCCheckBlock?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7907,32 +7935,19 @@ EpilogueVectorizerEpilogueLoop::createEpilogueVectorizedLoopSkeleton(
     EPI.SCEVSafetyCheck->getTerminator()->replaceUsesOfWith(
         VecEpilogueIterationCountCheck, LoopScalarPreHeader);
-  if (EPI.MemSafetyCheck)
+  if (EPI.MemSafetyCheck) {
```
undone, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7907,32 +7935,19 @@ EpilogueVectorizerEpilogueLoop::createEpilogueVectorizedLoopSkeleton(
-      VecEpilogueIterationCountCheck->getSinglePredecessor());
+  }
 
   DT->changeImmediateDominator(LoopScalarPreHeader,
```
Yep this one is still needed for now.

fhahn committed:
!fixup address latest comments, thanks!

fhahn committed:
[VPlan] Add PredIdx and SuccIdx arguments to connectBlocks (NFC).

fhahn committed:
[VPlan] Add insertOnEdge (NFC).

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn committed:
Merge branch 'main' into vplan-runtime-checks

fhahn committed:
!fixup use insertOnEdge.

fhahn committed:
[VPlan] Add insertOnEdge to VPBlockUtils (NFC).

Add a new helper to insert a new VPBlockBase on an edge between 2
blocks. Suggested in https://github.com/llvm/llvm-project/pull/114292
and also useful for some existing code.

fhahn committed:
Merge branch 'main' into vplan-runtime-checks

fhahn committed:
!fixup cleanup after merge.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn reviewed (commented):
The latest version now also updated the VPlan printing tests and moved final printing to VPlan::execute to also capture the updated skeleton

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn committed:
!fixup formatting

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1276,6 +1287,19 @@ VPlan *VPlan::duplicate() {
+    // Order is strict: first is the exit block, second is the scalar preheader.
+    return cast<VPBasicBlock>(MiddleVPBB->getSuccessors()[1]);
+  }
+  if (auto *IRVPBB = dyn_cast<VPBasicBlock>(MiddleVPBB->getSingleSuccessor()))
```
dyn_casting to VPBB but calling it `IRVPBB`?
Intention is to return the single successor if it's not an IRVPBB, as that represents the exit block?
```suggestion
  if (isa<VPIRBasicBlock>(MiddleVPBB->getSingleSuccessor())) {
    // Exit block is IRVPBB, scalar preheader is not.
    return nullptr;
  }
  return cast<VPBasicBlock>(MiddleVPBB->getSingleSuccessor());
```

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1276,6 +1287,19 @@ VPlan *VPlan::duplicate() {
+  auto *MiddleVPBB =
+      cast<VPBasicBlock>(getVectorLoopRegion()->getSingleSuccessor());
+  if (MiddleVPBB->getNumSuccessors() == 2) {
+    // Order is strict: first is the exit block, second is the scalar preheader.
```
Simpler to return last successor, provided it ain't a VPIRBB?

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1410,7 +1410,7 @@ class VPIRInstruction : public VPRecipeBase {
                               VPCostContext &Ctx) const override;
 
-  Instruction &getInstruction() { return I; }
+  Instruction &getInstruction() const { return I; }
```
Independent?

fhahn replied:
Done in 466ff3ed70e50343c9617de318e1b40efa03529d, thanks!

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn commented on llvm/test/Transforms/LoopVectorize/vplan-predicate-switch.ll:
```diff
@@ -76,25 +79,26 @@ define void @switch4_default_common_dest_with_case(ptr %start, ptr %end) {
 ; CHECK-EMPTY:
-; CHECK-NEXT: scalar.ph:
+; CHECK-NEXT: ir-bb<scalar.ph>:
+; CHECK-NEXT:   IR %bc.resume.val = phi ptr [ %ind.end, %middle.block ], [ %start, %entry ]
```
We now check the VPIRBB instead of the regular VPBB

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2429,6 +2429,30 @@ InnerLoopVectorizer::getOrCreateVectorTripCount(BasicBlock *InsertBlock) {
+  VPBlockBase *ScalarPH = Plan.getScalarPreheader();
+  VPBlockBase *PredVPB = Plan.getEntry();
+  VPBlockUtils::connectBlocks(PredVPB, ScalarPH, -1, 0);
+  VPBlockUtils::connectBlocks(PredVPB, VectorPH, 0, -1);
```
Updated to connect + swap, thanks.

> Furthermore, TCCheckBlock could then be inserted on this Entry-->vector_PH edge, so that connectScalarPreheaderInVPlan() assimilates introduceCheckBlockInVPlan().

Not sure how exactly, left as is for now.

fhahn commented on llvm/unittests/Transforms/Vectorize/VPlanVerifierTest.cpp:
```diff
@@ -233,8 +238,9 @@ TEST(VPVerifierTest, BlockOutsideRegionWithParent) {
+  VPBlockUtils::connectBlocks(VPPH, R1);
   VPBlockUtils::connectBlocks(R1, ScalarHeaderVPBB);
-  VPlan Plan(VPPH, &*TC, VPBB1, ScalarHeaderVPBB);
+  VPlan Plan(VPPH, &*TC, ScalarHeaderVPBB);
```
Done, thanks!

fhahn committed:
!fixup address latest comments, thanks!

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2429,6 +2429,30 @@ InnerLoopVectorizer::getOrCreateVectorTripCount(BasicBlock *InsertBlock) {
+static void introduceCheckBlockInVPlan(VPlan &Plan, BasicBlock *CheckIRBB) {
+  VPBlockBase *ScalarPH = Plan.getScalarPreheader();
+  VPBlockBase *VectorPH = Plan.getVectorPreheader();
+  VPBlockBase *PreVectorPH = VectorPH->getSinglePredecessor();
```
Added an assert for now, than

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -3237,13 +3237,17 @@ void VPWidenPointerInductionRecipe::print(raw_ostream &O, const Twine &Indent,
+  if (State.ExpandedSCEVs.contains(Expr)) {
+    State.Builder.SetInsertPoint(State.CFG.VPBB2IRBB[getParent()]);
+    return;
+  }
```
Added a comment, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -195,15 +195,6 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
-#endif
-      errs() << "not in a VPIRBasicBlock!\n";
-      return false;
-    }
```
We now can have regular recipes in VPIRBBs ( e.g expand-scev), adjust the check.

fhahn committed:
!fixup update verifier

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vplan-vp-select-intrinsics.ll:
```diff
@@ -7,11 +7,22 @@
+ ; IF-EVL-NEXT: Successor(s): ir-bb<scalar.ph>, ir-bb<vector.ph>
+
+ ; IF-EVL:      ir-bb<vector.ph>:
+ ; IF-EVL-NEXT:   IR   %4 = call i64 @llvm.vscale.i64()
+ ; IF-EVL-NEXT:   IR   %5 = mul i64 %4, 4
+ ; IF-EVL-NEXT:   IR   %6 = sub i64 %5, 1
+ ; IF-EVL-NEXT:   IR   %n.rnd.up = add i64 %N, %6
+ ; IF-EVL-NEXT:   IR   %n.mod.vf = urem i64 %n.rnd.up, %5
+ ; IF-EVL-NEXT:   IR   %n.vec = sub i64 %n.rnd.up, %n.mod.vf
+ ; IF-EVL-NEXT:   IR   %7 = call i64 @llvm.vscale.i64()
+ ; IF-EVL-NEXT:   IR   %8 = mul i64 %7, 4
```
we are now matching the VPIRBB

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1266,6 +1293,15 @@ VPlan *VPlan::duplicate() {
+  auto *MiddleVPBB =
+      cast<VPBasicBlock>(getVectorLoopRegion()->getSingleSuccessor());
+  auto *LastSucc = MiddleVPBB->getSuccessors().back();
+  // Order is strict: if the last successor is VPIRBasicBlock, it must be the
+  // single exit.
```
```suggestion
  // If scalar preheader is connected to VPlan, it is the last successor of MiddleVPBB.
  // If this last successor is a VPIRBasicBlock, it is the Exit block rather than the scalar preheader.
```

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -823,16 +821,25 @@ void VPRegionBlock::print(raw_ostream &O, const Twine &Indent,
 }
 #endif
 
+VPlan::VPlan(VPBasicBlock *Preheader, VPValue *TC, VPBasicBlock *Entry,
```
Perhaps some renaming can help be more consistent yet support gradual transition, e.g.,
```suggestion
VPlan::VPlan(VPBasicBlock *OriginalPreheader, VPValue *TC, VPBasicBlock *EntryVectorPreHeader,
```
?

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn committed:
!fixup address comments, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3762,36 +3757,27 @@ class VPlan {
 
   ~VPlan();
 
+  void setEntry(VPBasicBlock *VPBB) {
+    Entry = VPBB;
+    VPBB->setPlan(this);
+  }
```
Unfortunately I don't think so at the moment. All VPlans were created with the original preheader of the loop wrapped in a VPIRBB on construction.

After executing the main VPlan, we need to update the epilogue VPlan's entry to have a different VPIRBB, otherwise we would try to connect the original entry to the epilogue loop.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -855,9 +862,13 @@ VPlanPtr VPlan::createInitialVPlan(Type *InductionTy,
   VPIRBasicBlock *Entry =
       VPIRBasicBlock::fromBasicBlock(TheLoop->getLoopPreheader());
   VPBasicBlock *VecPreheader = new VPBasicBlock("vector.ph");
+  // Connect entry only to vector preheader initially. Edges to the scalar
+  // preheader will be inserted later, during skeleton creation when runtime
```
```suggestion
  // Connect entry only to vector preheader initially. Entry will also be connected to the scalar
  // preheader later, during skeleton creation when runtime
```

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1005,16 +1016,23 @@ void VPlan::execute(VPTransformState *State) {
   BasicBlock *MiddleBB = State->CFG.ExitBB;
-  VPBasicBlock *MiddleVPBB = getMiddleBlock();
   BasicBlock *ScalarPh = MiddleBB->getSingleSuccessor();
   replaceVPBBWithIRVPBB(getScalarPreheader(), ScalarPh);
+  VPBasicBlock *MiddleVPBB = getMiddleBlock();
   replaceVPBBWithIRVPBB(MiddleVPBB, MiddleBB);
```
```suggestion
  replaceVPBBWithIRVPBB(getVectorPreheader(), VectorPreHeader);
  replaceVPBBWithIRVPBB(getMiddleBlock(), MiddleBB);
  replaceVPBBWithIRVPBB(getScalarPreheader(), ScalarPh);
```
following the logical order listed in the comment? MiddleVPBB can be inlined for consistency.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -1005,16 +1016,23 @@ void VPlan::execute(VPTransformState *State) {
+  // blocks are created during skeleton creation, so we can only create the
+  // VPIRBasicBlocks now during VPlan execution rather than earlier during VPlan
+  // construction.
+  replaceVPBBWithIRVPBB(getVectorPreheader(), VectorPreHeader);
```
```suggestion
```
better positioned below along with the other replaceVPBBWithIRVPBB(); in any case better deal with them according to their natural order.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4177,10 +4176,6 @@ class VPBlockUtils {
                             unsigned PredIdx = -1u, unsigned SuccIdx = -1u) {
     assert((From->getParent() == To->getParent()) &&
            "Can't connect two block with different parents");
-    assert((SuccIdx != -1u || From->getNumSuccessors() < 2) &&
-           "Blocks can't have more than two successors.");
-    assert((PredIdx != -1u || To->getNumPredecessors() < 2) &&
-           "Blocks can't have more than two predecessors.");
```
These asserts no longer hold?

fhahn replied:
The second one doesn't hold any more, e.g. scalar.ph can now have more than 2 predecessors. Restored the first one.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -3235,13 +3235,21 @@ void VPWidenPointerInductionRecipe::print(raw_ostream &O, const Twine &Indent,
+
   const DataLayout &DL = State.CFG.PrevBB->getDataLayout();
   SCEVExpander Exp(SE, DL, "induction");
-
```
(ok, unrelated)

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -194,6 +194,7 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
       errs() << "not in a VPIRBasicBlock!\n";
       return false;
     }
+
```
(ok, unrelated)

ayalz commented on llvm/test/Transforms/LoopVectorize/RISCV/riscv-vector-reverse.ll:
```diff
@@ -142,59 +142,8 @@ define void @vector_reverse_i64(ptr nocapture noundef writeonly %A, ptr nocaptur
-; CHECK-NEXT:  Successor(s): vector loop
-; CHECK-EMPTY:
-; CHECK-NEXT:  <x1> vector loop: {
-; CHECK-NEXT:    vector.body:
-; CHECK-NEXT:      EMIT vp<[[CAN_IV:%.+]]> = CANONICAL-INDUCTION
-; CHECK-NEXT:      vp<[[DEV_IV:%.+]]> = DERIVED-IV ir<%n> + vp<[[CAN_IV]]> * ir<-1>
-; CHECK-NEXT:      vp<[[STEPS:%.+]]> = SCALAR-STEPS vp<[[DEV_IV]]>, ir<-1>
-; CHECK-NEXT:      CLONE ir<%i.0> = add nsw vp<[[STEPS]]>, ir<-1>
-; CHECK-NEXT:      CLONE ir<%idxprom> = zext ir<%i.0>
-; CHECK-NEXT:      CLONE ir<%arrayidx> = getelementptr inbounds ir<%B>, ir<%idxprom>
-; CHECK-NEXT:      vp<[[VEC_PTR:%.+]]> = reverse-vector-pointer inbounds ir<%arrayidx>, vp<[[VF]]>
-; CHECK-NEXT:      WIDEN ir<%13> = load vp<[[VEC_PTR]]>
-; CHECK-NEXT:      WIDEN ir<%add9> = add ir<%13>, ir<1>
-; CHECK-NEXT:      CLONE ir<%arrayidx3> = getelementptr inbounds ir<%A>, ir<%idxprom>
-; CHECK-NEXT:      vp<[[VEC_PTR2:%.+]]> = reverse-vector-pointer inbounds ir<%arrayidx3>, vp<[[VF]]>
-; CHECK-NEXT:      WIDEN store vp<[[VEC_PTR2]]>, ir<%add9>
-; CHECK-NEXT:      EMIT vp<[[CAN_IV_NEXT:%.+]]> = add nuw vp<[[CAN_IV]]>, vp<[[VFxUF]]>
-; CHECK-NEXT:      EMIT branch-on-count vp<[[CAN_IV_NEXT]]>, vp<[[VEC_TC]]>
-; CHECK-NEXT:    No successors
-; CHECK-NEXT:  }
-; CHECK-NEXT:  Successor(s): middle.block
-; CHECK-EMPTY:
-; CHECK-NEXT:  middle.block:
-; CHECK-NEXT:    EMIT vp<[[CMP:%.+]]> = icmp eq vp<[[TC]]>, vp<[[VEC_TC]]>
-; CHECK-NEXT:    EMIT branch-on-cond vp<[[CMP]]>
-; CHECK-NEXT:  Successor(s): ir-bb<for.cond.cleanup.loopexit>, scalar.ph
-; CHECK-EMPTY:
-; CHECK-NEXT:  ir-bb<for.cond.cleanup.loopexit>:
-; CHECK-NEXT:  No successors
-; CHECK-EMPTY:
-; CHECK-NEXT:  scalar.ph:
-; CHECK-NEXT:  Successor(s): ir-bb<for.body>
-; CHECK-EMPTY:
-; CHECK-NEXT:  ir-bb<for.body>:
-; CHECK-NEXT:    IR   %indvars.iv = phi i64 [ %0, %for.body.preheader ], [ %indvars.iv.next, %for.body ]
-; CHECK-NEXT:    IR   %i.0.in8 = phi i32 [ %n, %for.body.preheader ], [ %i.0, %for.body ]
-; CHECK:         IR   %indvars.iv.next = add nsw i64 %indvars.iv, -1
-; CHECK-NEXT:  No successors
-; CHECK-NEXT:  }
-; CHECK:  LV: Loop does not require scalar epilogue
```
Why/Should these checks be removed?

fhahn replied:
That was an accident during a merge, updated, thanks!

ayalz commented on llvm/test/Transforms/LoopVectorize/RISCV/riscv-vector-reverse.ll:
```diff
@@ -356,59 +305,7 @@ define void @vector_reverse_f32(ptr nocapture noundef writeonly %A, ptr nocaptur
-; CHECK:         IR   %indvars.iv.next = add nsw i64 %indvars.iv, -1
-; CHECK-NEXT:  No successors
-; CHECK-NEXT:  }
-; CHECK:  LV: Loop does not require scalar epilogue
```
ditto

fhahn replied:
That was an accident during a merge, updated, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8074,6 +8090,15 @@ EpilogueVectorizerEpilogueLoop::emitMinimumVectorEpilogueIterCountCheck(
   }
   ReplaceInstWithInst(Insert->getTerminator(), &BI);
   LoopBypassBlocks.push_back(Insert);
+
```
All VPlans are initialized with the original scalar preheader as their Entry, but this works only when vectorizing the main loop (w/ or w/o vectorizing the epilog). When vectorizing the epilog loop, Entry no longer serves to host SCEV-expand recipes but should instead refer to the block between Middle and "Vector Epilog PreHeader" (called "Vector Epilogue Trip Count Check", but that's also how the first block is named), as depicted in https://llvm.org/docs/Vectorizers.html#epilogue-vectorization. The documentation of Entry in createInitialVPlan() deserves update?

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn committed:
!fixup address latest comments, thanks!

fhahn committed:
!fixup restore part of assert and fix formatting

fhahn committed:
!fixup more formatting fixes

ayalz reviewed (approved):
This LGTM, thanks!

Added few last minor comments; tests seems to also reflect the patch converting canonical and EVL header phi's to scalar header phi's(?).

fhahn committed:
[VPlan] Use RPOT for VPlan codegen and printing.

This split off changes for more complex CFGs in VPlan from both
    https://github.com/llvm/llvm-project/pull/114292
    https://github.com/llvm/llvm-project/pull/112138

This reduces their respective diffs.

fhahn committed:
Merge branch 'main' into vplan-runtime-checks

fhahn committed:
!fixup update test

fhahn committed:
!fixup update test

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn committed:
!Fixup update after merge

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn commented on llvm/test/Transforms/LoopVectorize/AArch64/extractvalue-no-scalarization-required.ll:
```diff
@@ -64,6 +64,7 @@ exit:
 declare float @powf(float, float) readnone nounwind
 
 ; Ensure the extractvalue + add instructions are hoisted out
+; CM-LABEL: Checking a loop in 'test_getVectorCallCost'
```
Yes, the test was matching the VPlan from the previous function, split off to 0cedb8f215965fe25812094a9e2970b56d86a56e

fhahn committed:
!fixup fix formatting

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn committed:
!fixup use first successor for middle block.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-runtime-checks

fhahn committed:
!fixup update after merging #112138

fhahn committed:
!fixup use front instead of [0]

fhahn commented:
Made a minor update after  #112138: `getMiddleBlock` now uses `getScalarPreheader()->getPredecessors().front()` instead of `getScalarPreheader()->getSinglePredecessor()`, while before it was getting the middle block from the successor of the vector loop region
